File: lib/Commands.ml
open FOL
open LK
open Claire
open Env

exception CannotSolve of judgement list
exception FailedToApply
exception WrongArgument of argument
exception WrongArguments of argument list

let rec replicate i v =
  if i = 0 then []
  else v :: replicate (i-1) v

let findIndex f ls =
  let rec g = function
    | n,x::_ when f x-> Some n
    | n,_::xs -> g (n+1,xs)
    | _,_ -> None
  in
  g (0,ls)
let elemIndex e = findIndex(fun a-> e=a)

let onlyL : int -> int -> rule list =
  fun i n -> List.concat (replicate i [WL] @ replicate (n-i-1) [PL 1; WL])

let onlyR : int -> int -> rule list =
  fun i n -> List.concat (replicate i [WR] @ replicate (n-i-1) [PR 1; WR])

let assumption : env -> argument -> judgement list -> command list =
fun env ArgEmpty js ->
  match js with
  | Judgement(assms, props)::_ ->
    begin match findIndex (fun a -> List.mem a assms) props with
    | None -> raise (CannotSolve js)
    | Some i ->
      let Some j = elemIndex (List.nth props i) assms in
      [Apply(onlyR i (List.length props) @ onlyL j (List.length assms) @ [I])]
    end
  | _ -> raise (WrongArgument ArgEmpty)

(*| implyR

thm: a ==> b
goal: assms |- b, props

use thm
  assms, a ==> b |- b, props
apply ImpL
  assms |- a, b, props
  assms, b |- b, props
defer
  assms, b |- b, props
  assms |- a, b, props
assumption
  assms |- a, b, props
apply (PR 1, WR)
  assms |- a, props
*)
let rec implyR : env -> argument -> judgement list -> command list =
  fun env arg js -> match arg with
  | ArgIdents [(i,ps)] -> Use(i, ps) :: implyR env ArgEmpty js
  | ArgEmpty -> 
    [ Apply [ImpL];
      NewCommand("defer", ArgEmpty);
      NewCommand("assumption", ArgEmpty);
      Apply [PR 1; WR];
    ]
  | arg -> raise (WrongArgument arg)

(*| implyL

thm: a ==> b
goal: assms, a |- props

use thm
  assms, a, a ==> b |- props
apply ImpL
  assms, a |- a, props
  assms, a, b |- props
assumption
  assms, a, b |- props
apply (PL 1, WL)
  assms, b |- props
*)
let rec implyL : env -> argument -> judgement list -> command list =
  fun env arg js -> match arg with
  | ArgIdents [i,ps] -> Use(i, ps) :: implyL env ArgEmpty js
  | ArgEmpty -> [Apply[ImpL];NewCommand("assumption", ArgEmpty);Apply[PL 1; WL]]
  | arg -> raise (WrongArgument arg)

(*| genR

goal: assms |- P(a), props

apply Cut [Forall a. P(a)]
  assms |- Forall a. P(a), P(a), props
  assms, Forall a. P(a) |- P(a), props
defer
  assms, Forall a. P(a) |- P(a), props
  assms, |- Forall a. P(a), P(a), props
apply (ForallL [a])
  assms, P(a) |- P(a), props
  assms, |- Forall a. P(a), P(a), props
assumption
  assms, |- Forall a. P(a), P(a), props
apply (PR 1, WR)
  assms, |- Forall a. P(a), props
*)
let genR : env -> argument -> judgement list -> command list =
  fun env arg js -> match arg,js with
  | ArgIdents [(i,[])], Judgement(_, (p::_))::_ ->
    [ Apply [Cut(Forall(i, p))];
      NewCommand("defer", ArgEmpty);
      Apply [ForallL(Var i)];
      NewCommand("assumption", ArgEmpty);
      Apply [PR 1; WR];
    ]
  | arg, _ -> raise (WrongArgument arg)

(*| genL

goal: assms, P(a) |- props

apply Cut [Forall a. P(a)]
  assms, P(a) |- Forall a. P(a), props
  assms, P(a), Forall a. P(a) |- props
apply (ForallR [a])
  assms, P(a) |- P(a), props
  assms, P(a), Forall a. P(a) |- props
assumption
  assms, P(a), Forall a. P(a) |- props
apply (PL 1, WR)
  assms, Forall a. P(a) |- props
*)
let genL : env -> argument -> judgement list -> command list =
  fun env arg js -> match arg,js with
  | ArgIdents [(i,[])], Judgement(p::ps,_)::_ -> 
    [ Apply [Cut(Forall(i, p))];
      Apply [ForallR i];
      NewCommand("assumption", ArgEmpty);
      Apply [PL (List.length ps); WL];
    ]
  | arg, _ -> raise (WrongArgument arg)

(*| absR

goal: assms, a |- b, props

apply Cut [a ==> b]
  assms, a |- a ==> b, b, props
  assms, a, a ==> b |- b, props
defer
  assms, a, a ==> b |- b, props
  assms, a |- a ==> b, b, props
apply ImpL
  assms, a |- a, b, props
  assms, a, b |- b, props
  assms, a |- a ==> b, b, props
assumption [2]
  assms, a |- a ==> b, b, props
apply (PR 1, WR, WL)
  assms |- a ==> b, props
*)

let absL : env -> argument -> judgement list -> command list =
fun env arg js -> match arg,js with
  | ArgEmpty, Judgement(a::_, (b::_))::_ ->
    [ Apply [Cut(Then(a,b))];
      NewCommand("defer", ArgEmpty);
      Apply [ImpL];
      NewCommand("assumption", ArgEmpty);
      NewCommand("assumption", ArgEmpty);
      Apply [PR 1; WR; WL];
    ]
  | arg, _ -> raise (WrongArgument arg)

let export_command : (string * (env -> argument -> judgement list -> command list)) list = [
  "assumption", assumption;
  "implyR", implyR;
  "implyL", implyL;
  "genR", genR;
  "genL", genL;
  "absL", absL;
]

(* --------------------------------------------- *)

let definition : argument list -> decl list = function
  | [ArgTyped(i, typ); ArgPreds[PredFml body]] -> [ConstD(i, typ); AxiomD(i ^ "_def", body)]
  | arg -> Printf.printf "arg=%s\n" (show_arguments arg); raise (WrongArguments arg)
let export_decl = ["definition", definition]

File: lib/commands.pl
:- module('lib/commands',[]).

replicate(0,_,[]).
replicate(N,V,[V|Vs]) :- N1 is N - 1, replicate(N1,V,Vs).
findIndex(F,Ls,R) :- findIndex1(F,0,Ls,R).
findIndex1(F,N,[X|_],N) :- call(F,X),!.
findIndex1(F,N,[_|Xs],R) :- N1 is N + 1, findIndex1(F,N1,Xs,R).
elemIndex(E,Ls,R) :- findIndex(=(E),Ls,R).
onlyL(I,N,Rs) :-
  replicate(I,[wL],R1),NI1 is N-I-1,replicate(NI1,[pL(1),wL],R2),append(R1,R2,R3),append(R3,Rs).
onlyR(I,N,Rs) :-
  replicate(I,[wR],R1),NI1 is N-I-1,replicate(NI1,[pR(1),wR],R2),append(R1,R2,R3),append(R3,Rs).
assumption(_,[],[(Assms⊦Props)|_],[apply(Rs)]) :-
  findIndex([A]>>member(A,Assms),Props,I),!,
  nth0(I,Props,I2),elemIndex(I2,Assms,J),!,
  length(Props,Pi),length(Assms,Aj),
  onlyR(I,Pi,Ps),onlyL(J,Aj,As),append([Ps,As,[i]],Rs).
assumption(_,[],[(Assms⊦Props)|Js],_) :- throw(cannotSolve([(Assms⊦Props)|Js])).
assumption(_,_,_,_) :- throw(wrongArgument([])).
implyR(Env,i([(I,Ps)]),Js,[use(I, Ps)| Cs1]) :- implyR(Env,[],Js,Cs1).
implyR(_,[],_,
    [ apply([impL]),defer*[],assumption*[],apply([pR(1), wR])]) :- !.
implyR(_,Arg,_,_) :- throw(wrongArgument(Arg)).
implyL(Env,i([(I,Ps)]),Js,[use(I,Ps)|Cs]) :- implyL(Env,[],Js,Cs).
implyL(_,[],_,[apply([impL]),assumption*[],apply([pL(1),wL])]).
implyL(_,Arg,_,_) :- throw(wrongArgument(Arg)).
genR(_,i([(I,[])]),[_ ⊦ [P|_] |_],[
  apply([cut(forall(I, P))]),
  defer*[],
  apply([forallL(I)]),
  assumption*[],
  apply([pR(1), wR])
]) :- !.
genR(_,Arg,Js,_) :- throw(wrongArgument(Arg,Js)).
genL(_,i([(I,[])]),[[P|Ps] ⊦ _ |_],[
  apply([cut(forall(I, P))]),
  apply([forallR(I)]),
  assumption*[],
  apply([pL(PLen), wL])
]) :- length(Ps,PLen).
genL(_,Arg,_,_) :- throw(wrongArgument(Arg)).
absL(_,[],[[A|_] ⊦ [B|_]|_],[
  apply([cut(A==>B)]),
  defer*[],
  apply([impL]),
  assumption*[],
  assumption*[],
  apply([pR(1),wR, wL])
]).
absL(_,Arg,Js,_) :- throw(wrongArgument(Arg,Js)).

export_command([assumption,implyR,implyL,genR,genL,absL]).

/* --------------------------------------------- */

definition([n(I:Typ),p([Body])],[constant(I,Typ),axiom(I2,Body)]) :- format(atom(I2),'~w_def',[I]).
definition(Arg,_) :- throw(wrongArgument(Arg)).

export_decl([definition]).
